cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -923,29 +923,41 @@ private boolean isSwapViolatingContainerLimit(double sourceUtilizationDelta,
 
   /**
    * @param clusterModel the cluster topology and load.
+   * @param optimizationOptions Options to adjust balance upper limit in case goal optimization is triggered by goal
+   * violation detector.
    * @return the utilization upper threshold in percent for the {@link #resource()}
    */
-  private double computeBalanceUpperThreshold(ClusterModel clusterModel) {
+  private double computeBalanceUpperThreshold(ClusterModel clusterModel, OptimizationOptions optimizationOptions) {
     return (clusterModel.load().expectedUtilizationFor(resource()) / clusterModel.capacityFor(resource()))
-        * (1 + balancePercentageWithMargin(resource()));
+           * (1 + balancePercentageWithMargin(optimizationOptions));
   }
 
   /**
    * @param clusterModel the cluster topology and load.
+   * @param optimizationOptions Options to adjust balance lower limit in case goal optimization is triggered by goal
+   * violation detector.
    * @return the utilization lower threshold in percent for the {@link #resource()}
    */
-  private double computeBalanceLowerThreshold(ClusterModel clusterModel) {
+  private double computeBalanceLowerThreshold(ClusterModel clusterModel, OptimizationOptions optimizationOptions) {
     return (clusterModel.load().expectedUtilizationFor(resource()) / clusterModel.capacityFor(resource()))
-        * Math.max(0, (1 - balancePercentageWithMargin(resource())));
+           * Math.max(0, (1 - balancePercentageWithMargin(optimizationOptions)));
   }
 
   /**
    * To avoid churns, we add a balance margin to the user specified rebalance threshold. e.g. when user sets the
    * threshold to be resourceBalancePercentage, we use (resourceBalancePercentage-1)*balanceMargin instead.
+   *
+   * @param optimizationOptions Options to adjust balance percentage with margin in case goal optimization is triggered
+   * by goal violation detector.
    * @return the rebalance threshold with a margin.
    */
-  private double balancePercentageWithMargin(Resource resource) {
-    return (_balancingConstraint.resourceBalancePercentage(resource) - 1) * BALANCE_MARGIN;
+  private double balancePercentageWithMargin(OptimizationOptions optimizationOptions) {
+    double balancePercentage = optimizationOptions.isTriggeredByGoalViolation()
+                               ? _balancingConstraint.resourceBalancePercentage(resource())
+                                 * _balancingConstraint.goalViolationDistributionThresholdMultiplier()
+                               : _balancingConstraint.resourceBalancePercentage(resource());
```

Comment:
maybe we can change to
`_balancingConstraint.resourceBalancePercentage() * (optimizationOptions.isTriggeredByGoalViolation() ? _balancingConstraint.goalViolationDistributionThresholdMultiplier() : 1)`
to reduce a call to `_balancingConstraint.resourceBalancePercentage()`

Reply:
Replied this comment above.